### PR TITLE
fix(messaging,nnbd): regression in `RemoteMessage.fromMap()` causing silent failure

### LINKED ISSUE #4949: FirebaseMessaging: RemoteMessage.fromMap(): regression,  faulty handling of 'contentAvailable' & 'mutableContent'
## Bug report

Commit: 8a58dac3afbc2a80bfd18398158fdc0f75af1210    (part of PR #4909)

In RemoteMessage.fromMap():

Apparent regression in handling of missing 'contentAvailable' & 'mutableContent' message fields in message map.

These fields are now initialised to 'false' (and are therefore boolean). If the fields are not present in the map, the code attempts to set the value to 'null' (the return from map[] when the key is not present). 

This fails as null is not a subtype of bool, although the exception is seemingly not propagated to the application.

Earlier versions of this had, for example,

  contentAvailable: map['contentAvailable']??false,

which maintains the bool type. For whatever reason this commit removed the '??false'.

This results in silent failure to deliver messages to the package user (at least for messages where these optional; fields are not present.

### Expected behavior

Restore the '??false' to maintain type correctness.
Local experimentation shows that this seems to work.

I would write a PR but I'm not familiar enough with the structure and workflows of flutterfire to do it within any reasonable time

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel beta, 1.25.0-8.3.pre, on Microsoft Windows [Version 10.0.19041.746], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Chrome - develop for the web
[√] Android Studio (version 4.1.0)
[√] VS Code (version 1.52.1)      
[√] Connected device (3 available)

• No issues found!
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
Dart SDK 2.12.0-133.7.beta
Flutter SDK 1.25.0-8.3.pre
no_words 1.0.0-nullsafety.0

dependencies:
- cloud_firestore 0.17.0-1.0.nullsafety.0 [cloud_firestore_platform_interface cloud_firestore_web firebase_core firebase_core_platform_interface flutter meta]
- cloud_functions 0.9.1-1.0.nullsafety.1 [cloud_functions_platform_interface cloud_functions_web firebase_core firebase_core_platform_interface flutter]
- cupertino_icons 1.0.2
- firebase_auth 0.21.0-1.1.nullsafety.1 [firebase_auth_platform_interface firebase_auth_web firebase_core firebase_core_platform_interface flutter meta]
- firebase_core 0.8.0-1.0.nullsafety.1 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_messaging 9.0.0-1.0.nullsafety.0 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- freezed_annotation 0.13.0-nullsafety.0 [collection json_annotation meta]
- hooks_riverpod 0.13.0-nullsafety.1 [collection flutter flutter_hooks flutter_riverpod riverpod state_notifier]
- logging 1.0.0-nullsafety.0
- meta 1.3.0-nullsafety.6
- rxdart 0.26.0-nullsafety.1

dev dependencies:
- build_runner 1.11.1 [args async build build_config build_daemon build_resolvers build_runner_core code_builder collection crypto dart_style glob graphs http_multi_server io js logging meta mime path pedantic pool pub_semver pubspec_parse shelf shelf_web_socket stack_trace stream_transform timing watcher web_socket_channel yaml]
- effective_dart 1.3.0
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]
- freezed 0.13.0-nullsafety.2 [analyzer build build_config meta source_gen freezed_annotation]

transitive dependencies:
- _fe_analyzer_shared 14.0.0 [meta]
- analyzer 0.41.2 [_fe_analyzer_shared args cli_util collection convert crypto glob meta package_config path pub_semver source_span watcher yaml]
- args 1.6.0
- async 2.5.0-nullsafety.3 [collection]
- boolean_selector 2.1.0-nullsafety.3 [source_span string_scanner]
- build 1.6.2 [analyzer async convert crypto glob logging meta path]
- build_config 0.4.5 [checked_yaml json_annotation meta path pubspec_parse yaml]
- build_daemon 2.1.7 [built_collection built_value http_multi_server logging pedantic path pool shelf shelf_web_socket stream_transform watcher web_socket_channel]
- build_resolvers 1.5.3 [analyzer build crypto graphs logging meta path package_config pool pub_semver]
- build_runner_core 6.1.7 [async build build_config build_resolvers collection convert crypto glob graphs logging meta path package_config pedantic pool timing watcher yaml]
- built_collection 4.3.2 [collection quiver]
- built_value 7.1.0 [built_collection collection fixnum quiver]
- characters 1.1.0-nullsafety.5
- charcode 1.2.0-nullsafety.3
- checked_yaml 1.0.4 [json_annotation source_span yaml]
- cli_util 0.2.0 [path]
- clock 1.1.0-nullsafety.3
- cloud_firestore_platform_interface 4.0.0-1.0.nullsafety.0 [collection firebase_core flutter meta plugin_platform_interface]  
- cloud_firestore_web 0.4.0-1.0.nullsafety.0 [cloud_firestore_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- cloud_functions_platform_interface 4.0.2-1.0.nullsafety.1 [firebase_core flutter meta plugin_platform_interface]
- cloud_functions_web 3.1.4-1.0.nullsafety.1 [cloud_functions_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- code_builder 3.6.0 [built_collection built_value collection matcher meta]
- collection 1.15.0-nullsafety.5
- convert 2.1.1 [charcode typed_data]
- crypto 2.1.5 [collection convert typed_data]
- dart_style 1.3.10 [analyzer args path source_span]
- fake_async 1.2.0-nullsafety.3 [clock collection]
- firebase_auth_platform_interface 4.0.0-1.1.nullsafety.1 [firebase_core flutter meta plugin_platform_interface]
- firebase_auth_web 0.4.0-1.1.nullsafety.1 [firebase_auth_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins http_parser intl js meta]
- firebase_core_platform_interface 4.0.0-1.0.nullsafety.1 [flutter meta plugin_platform_interface]
- firebase_core_web 0.3.0-1.0.nullsafety.1 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_messaging_platform_interface 2.0.0-1.0.nullsafety.0 [firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 0.2.0-1.0.nullsafety.0 [firebase_core firebase_core_web firebase_messaging_platform_interface flutter 
flutter_web_plugins js meta]
- fixnum 0.10.11
- flutter_hooks 0.16.0-nullsafety.0 [flutter]
- flutter_riverpod 0.13.0-nullsafety.1 [collection flutter meta riverpod state_notifier]
- flutter_web_plugins 0.0.0 [flutter js characters collection meta typed_data vector_math]
- glob 1.2.0 [async collection node_io path pedantic string_scanner]
- graphs 0.2.0
- http_multi_server 2.2.0 [async]
- http_parser 4.0.0-nullsafety [charcode collection source_span string_scanner typed_data]
- intl 0.17.0-nullsafety.2 [clock path]
- io 0.3.4 [charcode meta path string_scanner]
- js 0.6.3-nullsafety.3
- json_annotation 4.0.0-nullsafety.0
- matcher 0.12.10-nullsafety.3 [stack_trace]
- mime 0.9.7
- node_interop 1.2.1 [js]
- node_io 1.1.1 [node_interop path]
- package_config 1.9.3 [path charcode]
- path 1.8.0-nullsafety.3
- pedantic 1.10.0-nullsafety.3
- plugin_platform_interface 1.1.0-nullsafety.2 [meta]
- pool 1.5.0-nullsafety.3 [async stack_trace]
- pub_semver 1.4.4 [collection]
- pubspec_parse 0.1.7 [checked_yaml json_annotation pub_semver yaml]
- quiver 2.1.5 [matcher meta]
- riverpod 0.13.0-nullsafety.1 [collection freezed_annotation meta state_notifier]
- shelf 1.0.0-nullsafety.0 [async collection http_parser path stack_trace stream_channel]
- shelf_web_socket 0.2.4 [shelf stream_channel web_socket_channel]
- sky_engine 0.0.99
- source_gen 0.9.10+1 [analyzer async build dart_style glob meta path pedantic source_span]
- source_span 1.8.0-nullsafety.4 [charcode collection path term_glyph]
- stack_trace 1.10.0-nullsafety.6 [path]
- state_notifier 0.7.0-nullsafety.0 [meta]
- stream_channel 2.1.0-nullsafety.3 [async]
- stream_transform 1.2.0
- string_scanner 1.1.0-nullsafety.3 [charcode source_span]
- term_glyph 1.2.0-nullsafety.3
- test_api 0.2.19-nullsafety.6 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- timing 0.1.1+3 [json_annotation]
- typed_data 1.3.0-nullsafety.5 [collection]
- vector_math 2.1.0-nullsafety.5
- watcher 0.9.7+15 [async path pedantic]
- web_socket_channel 1.1.0 [async crypto stream_channel]
- yaml 2.2.1 [charcode collection string_scanner source_span]
```

</details>

---

### PR DESCRIPTION
**Pull request by spacejunky**
## Description

Proposed correction of  Issue #4949, which causes silent discard of incoming FCM's when running under sound null safety.

This not a breaking change but a correction of a previous regression.

Due to multiple problems getting Melos to run on my windows system, despite attempting to follow the Contributor Guide, I have used the firebase_messaging example app to provide a demo of the faulty behavior and its subsequent correction. This is neither unit or integration testing. Sorry!

The changes made to the example app so that it runs under sound null safety (c1fda5b) -- which is when the faulty behavior shows up -- are not the result of any deep thought and are not proposed for merging.  The analysis & formatting fails are all to be found in this commit, I believe.

Likewise not tracking.md (f5cd8e5), which just provides commentary for your convenience. 

Only the two lines in remote_message.dart, commit 48bc692, are relevant to the issue. I'm sorry  that I don't know how to formulate the PR so that it says only that, while still giving you the visibility of the supporting work.

## Related Issues

fix #4949 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [X] I read and followed the [Flutter Style Guide].
- [X] I signed the [CLA].
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by spacejunky** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here with `@googlebot I signed it!` and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2FFirebaseExtended%2Fflutterfire%2Fpull%2F5336) for more info**.

@googlebot I signed it!

**Comment by spacejunky** [comment]
I suspect Issue #5324 of being an occurrence of this as well

**Comment by GarvMaggu** [comment]
@spacejunky Hello, my background and foreground message handlers, both were not working on Flutter 2.0.1 with null safety on. Using this fix, it is working now. Thank you.

File: packages/firebase_messaging/firebase_messaging/example/lib/message.dart [hunk 1]
```diff
@@ -31,13 +30,13 @@ class MessageView extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
-    final MessageArguments args = ModalRoute.of(context).settings.arguments;
+    final MessageArguments args = ModalRoute.of(context)?.settings.arguments as MessageArguments;
```
Comment (rrousselGit) [inline packages/firebase_messaging/firebase_messaging/example/lib/message.dart:33]:
```suggestion
    final MessageArguments? args = ModalRoute.of(context)?.settings.arguments;
```

File: packages/firebase_messaging/firebase_messaging/example/lib/message.dart [hunk 2]
```diff
@@ -78,47 +77,47 @@ class MessageView extends StatelessWidget {
                   ),
                   row(
                     'Channel ID',
-                    notification.android.channelId,
+                    notification.android!.channelId,
```
Comment (rrousselGit) [inline packages/firebase_messaging/firebase_messaging/example/lib/message.dart:80]: `notification` is nullable but I don't see a null check for it, how is that possible?

File: packages/firebase_messaging/firebase_messaging/example/lib/tracking.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
+# Work on demo and fix for Issue #4949
```
Comment (rrousselGit) [inline packages/firebase_messaging/firebase_messaging/example/lib/tracking.md:1]: Is this meant to be deleted?

**Comment by rrousselGit** [comment]
Thanks!

Could you add tests for this?
Also the analyzer warnings should be fixed

File: packages/firebase_messaging/firebase_messaging/example/lib/message.dart (hunk 1 above)
Comment (spacejunky) [reply packages/firebase_messaging/firebase_messaging/example/lib/message.dart:33]: @rrousselGit : I wouldn't bother with all the stuff I did in the example app. As I said in the PR text, the only point of that was to get the official example far enough that I could use it to demonstrate that the original problem in 4949 also occurred in the example. I made no effort, to evaluate the example app changes enough to decide if they make sense or not.

The only useful thing in this PR are the two one liners in remote_message.dart. 
Everything else is only there for supporting info 

Sorry if that is confusing ...

File: packages/firebase_messaging/firebase_messaging/example/lib/tracking.md (hunk 3 above)
Comment (spacejunky) [reply packages/firebase_messaging/firebase_messaging/example/lib/tracking.md:1]: This just so you can see what I did to demo & test the changes.

Delete

**Comment by spacejunky** [comment]
> Could you add tests for this?

Done

> Also the analyzer warnings should be fixed

Analyzer warnings are all in the "rough" coversion of the example app to null-safety, which is not intended (and not suitable, for many reasons) for merging.

Cheers,

**Comment by rrousselGit** [comment]
> Analyzer warnings are all in the "rough" coversion of the example app to null-safety, which is not intended (and not suitable, for many reasons) for merging.

If this is not intended for merging, could you revert the changes? At they cause the CI to fail.

File: packages/firebase_messaging/firebase_messaging/example/test/remote_message_test.dart [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import 'package:firebase_messaging/firebase_messaging.dart';
```
Comment (Salakar) [inline packages/firebase_messaging/firebase_messaging/example/test/remote_message_test.dart:1]: Can you move this file into https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging_platform_interface/test 

Added tests should reside in the same package, you should just be able to move this into the directory above, though you'll need to adjust your import to use RemoteMessage from `firebase_messaging_platform_interface`

**Comment by spacejunky** [comment]
Moved as requested, noting that I am unable to verify correctness as (due to the multiple melos problems I have encounted) nothing in that part of the tree has will even compile on my system as the package paths are broken.

PS: Is melos supposed to work on Windows systems?

**Review by rrousselGit — APPROVED** [review]

File: packages/firebase_messaging/firebase_messaging_platform_interface/test/remote_message_test.dart [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import 'package:firebase_messaging_platfprm/firebase_messaging_platform.dart';
```
Comment (rrousselGit) [inline packages/firebase_messaging/firebase_messaging_platform_interface/test/remote_message_test.dart:1]:
```suggestion
import 'package:firebase_messaging_platform/firebase_messaging_platform.dart';
```

File: packages/firebase_messaging/firebase_messaging_platform_interface/test/remote_message_test.dart [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import 'package:firebase_messaging_platform/firebase_messaging_platform.dart';
```
Comment (Salakar) [inline packages/firebase_messaging/firebase_messaging_platform_interface/test/remote_message_test.dart:1]:
```suggestion
import 'package:firebase_messaging_platform_interface/firebase_messaging_platform_interface.dart';
```

**Review by Salakar — APPROVED** [review]

**Comment by pengw00** [comment]
@Salakar  hello, is it built to fix the IOS silent issue? haven't seen version updated for this PR.

**Pull request merged**